api/v1alpha7/types.go

```diff
@@ -163,20 +163,81 @@ type RootVolume struct {
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
 }
 
-type AdditionalBlockDevice struct {
-	// Name of the Cinder volume in the context of a machine.
-	// It will be combined with the machine name to make the actual volume name.
-	Name string `json:"name"`
-	// Size is the size in GB of the volume.
-	Size int `json:"diskSize"`
-	// VolumeType is the volume type of the volume.
-	// If omitted, the default type will be used.
-	VolumeType string `json:"volumeType,omitempty"`
-	// AvailabilityZone is the volume availability zone to create the volume in.
+// blockDeviceStorage is the storage type of a block device to create and
+// contains additional storage options.
+// +union
+//
+//nolint:godot
+type BlockDeviceStorage struct {
```

Comment:
You should add CEL to validate that the `Volume` field is present iff `type: Volume`. There's examples of this in the openshift API examples API group

Reply:
@JoelSpeed can you confirm that I in fact don't need it because the 2 fields in `BlockDeviceVolume` are optional?

Reply:
`BlockDeviceStorage` is a discriminated union, you need to apply the validations for it right now because the discriminated union code within the API is not yet implemented.

See https://github.com/openshift/api/blob/79b9cd5e6020c2e430b0d90ad7413a0ab85f55c8/example/v1/types_stable.go#L89 for an example.

To be clear, when `type: Local`, we do not want `volume` (the member) to be set, and when `type: Volume` we do want `volume` to be set right?

Reply:
> To be clear, when type: Local, we do not want volume (the member) to be set

yes

> when type: Volume we do want volume to be set right?

Not necessarily, both options in `BlockDeviceVolume` are optional and have no default; they are up to the OpenStack cloud and how it was deployed.

Reply:
IIUC I need one CEL validation for `type: local` that verifies that no `volume` is given.